crates/shared/src/ethrpc/multicall.rs

```diff
@@ -0,0 +1,426 @@
+//! Multicall encoding and decoding.
+
+use crate::ethrpc::dummy;
+use contracts::support::Multicall;
+use ethcontract::{
+    errors::ExecutionError,
+    tokens::{self, Tokenize as _},
+};
+use hex_literal::hex;
+use lazy_static::lazy_static;
+use std::iter;
+use web3::{
+    self,
+    api::Eth,
+    ethabi::{self, ParamType, Token},
+    types::{AccessList, BlockId, Bytes, CallRequest, H160, U256, U64},
+    Transport,
+};
+
+/// A single call in a multicall batch.
+#[derive(Clone, Debug, Default, Eq, PartialEq)]
+pub struct Call {
+    /// The address to call.
+    pub to: H160,
+
+    /// Optional gas limit to use for the call.
+    pub gas: Option<U256>,
+
+    /// Optional value to use for the call.
+    pub value: Option<U256>,
+
+    /// Data to use for the call.
+    pub data: Bytes,
+}
+
+impl Call {
+    /// Encode into a tuple used for ABI encoding.
+    pub fn encode(self) -> (H160, U256, U256, ethcontract::Bytes<Vec<u8>>) {
+        (
+            self.to,
+            self.gas.unwrap_or_default(),
+            self.value.unwrap_or_default(),
+            ethcontract::Bytes(self.data.0),
+        )
+    }
+}
+
+/// Additional options that affect all calls in the multicall.
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct Options {
+    /// The transaction origin.
+    ///
+    /// Note that this will not be the `msg.sender` for the calls, which will
+    /// instead be the Multicall trampoline contract.
+    pub origin: Option<H160>,
+
+    /// Total gas limit for all the calls.
+    pub gas: Option<U256>,
+
+    /// Gas price.
+    pub gas_price: Option<U256>,
+
+    /// The transaction type index.
+    pub transaction_type: Option<U64>,
+
+    /// Access list.
+    pub access_list: Option<AccessList>,
+
+    /// Max fee per gas.
+    pub max_fee_per_gas: Option<U256>,
+
+    /// Miner bribe.
+    pub max_priority_fee_per_gas: Option<U256>,
+}
+
+/// Web3 convenience multicall extension.
+#[async_trait::async_trait]
+pub trait MulticallExt<T> {
```

Comment:
> Why not just use a free method?

This is actually an oversight on my behalf. Specifically, the way it will end up being used requires these to be floating functions that don't actually execute against a `web3` instance (since they will be added to a buffer of JSON RPC request objects within a `web3::Transport` implementation) - so this will change a bit.

I can change it if you feel strongly, but it will likely not survive the next few PRs.

> Pros of using an `Ext` trait:

This is what lured me into using it like this. I thought the following symmetry looked "pretty" to me:
```rust
web3.eth().call(CallResult { ... });
web3.eth().multicall(vec![Call { ... }, Call { ... }]);
```